concurrency/DurableLocalStoreManagerImpl.java

```diff
@@ -0,0 +1,287 @@
+ * @author ssubra
+ *
+ */
+public class DurableLocalStoreManagerImpl implements DurableLocalStoreManager {
+    private static final S3Logger logger = new BigBirdFatalInterceptor();
+
+    protected final StorageInstanceManager sim;
+    
+    // bucketName -> DLS for that bucket
+    private final Map<String, DurableLocalStore> bucketDataMap;
+
+    /**
+     * This checks if the files are present for apply/export
+     * 
+     * @param dbDir
+     * @param bucketName
+     * @return
+     */
+    public static boolean verifyStoreFilesPresent(String dbDir, String bucketName) {
+        return DurableLocalJsonStore.verifyStoreFilesPresent(dbDir, bucketName);
+    }
+    
+    public static boolean verifyIntegrity(String dbDir, String bucketName) {
+        //TODO Add an implementation when we add the Verify task at NCListener side.
+        return false;
+    }
+
+    /**
+     * This is used to move the store files after an import - from the import location to actual.
+     */
+    public static void moveStoreFiles(String dbDir,
+                                      String newBucketName,
+                                      String fromDir,
+                                      String oldBucketName) throws Exception
+    {
+        DurableLocalJsonStore.moveStoreFiles(dbDir, newBucketName, fromDir, oldBucketName);
+    }
+
+    public static void dropStoreFiles(String dir, String bucket) throws IOException {
+        DurableLocalJsonStore.dropStoreFiles(dir, bucket);
+    }
+
+    /**
+     * Factory method to allow for a quick swap between stores. protected to allow for unit test overrides.
+     * 
+     * @return
+     */
+    protected DurableLocalStore newStore(String bucketName, boolean create) throws StorageNodeException {
+        return new DurableLocalJsonStore(sim, bucketName, create);
+    }
+
+    public DurableLocalStoreManagerImpl(StorageInstanceManager sim) {
+        this.sim = sim;
+        this.bucketDataMap = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public boolean putOrUpdate(String bucketName, IndexKey key, Item value) throws StorageNodeException {
+        final DurableLocalStore store = getBucketData(bucketName);
+        store.assertIfNotOpen("putOrUpdate");
+        return store.putOrUpdate(key, value);
+    }
+
+    @Override
+    public Item get(String bucketName, IndexKey key) throws StorageNodeException {
+        final DurableLocalStore store = getBucketData(bucketName);
+        store.assertIfNotOpen("get");
+        return store.get(key);
+    }
+
+    @Override
+    public void create(String bucketName) throws StorageNodeException {
+        DurableLocalStore dataStore = bucketDataMap.get(bucketName);
+
+        if (dataStore == null) {
+            synchronized (this) {
+                dataStore = bucketDataMap.get(bucketName);
+                if (dataStore == null) {
+                    dataStore = newStore(bucketName, true);
+                    bucketDataMap.put(bucketName, dataStore);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **112 and 115**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `put()` with `putIfAbsent()` and then removing the synchronization. The check is at line: **113**.

**More info**  
 `putIfAbsent()` is thread-safe and puts the value only if the `ConcurrentHashMap` does not contain the key. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java#L927-L932) (external link).